cert/spec.go

```diff
@@ -438,56 +365,114 @@ func (spec *Spec) ResetBackoff() {
 	spec.tr.Backoff.Reset()
 }
 
-// EnforcePKI processes a spec, updating content on disk, taking action as needed.
-// Returns (TTL for PKI, error).  If an error occurs, the ttl is at best
-// a hint to the invoker as to when the next refresh is required- that said
-// the invoker should back off and try a refresh.
-func (spec *Spec) EnforcePKI(enableActions bool) (time.Duration, error) {
-	err := spec.CheckDiskPKI()
+// checkDiskCertKey performs sanity checks against the cert/key read from disk, identifying
+// if it's valid and still usable.
+func (spec *Spec) checkDiskCertKey(ca *x509.Certificate) error {
+	existingCert, err := spec.Cert.ReadCertificate()
 	if err != nil {
-		log.Debugf("manager: %s, checkdiskpki: %s.  Forcing refresh.", spec, err.Error())
-		spec.ForceRenewal()
+		log.Debugf("spec %s: cert failed to be read: %s", spec, err)
+		return err
 	}
+	// update our internal time tracking while we're in here; even if immediately discard it,
+	// keeping it accurate to when we last saw it is desirable for metrics.
+	spec.updateCertExpiry(existingCert.NotAfter)
 
-	if err = spec.CheckCA(); err != nil {
-		log.Errorf("manager: the CA for %s has changed, but the service couldn't be notified of the change", spec)
+	keyData, err := spec.Key.ReadFile()
+	if err != nil {
+		log.Debugf("spec %s: key failed to be read: %s", spec, err)
+		return err
+	}
+	err = verifyCertChain(ca, existingCert)
+	if err != nil {
+		log.Debugf("spec %s: CA has changed, cert is no longer valid via it: %s", spec, err)
+		return err
+	}
+	err = spec.checkDiskPKI(existingCert, keyData)
+	if err != nil {
+		return err
+	}
+	now := time.Now().Add(spec.tr.Before)
+	if now.After(existingCert.NotAfter) {
+		return fmt.Errorf("certificate already expired at %s", existingCert.NotAfter)
 	}
+	if existingCert.NotBefore.After(now) {
+		// someone needs a better clock.
+		return fmt.Errorf("certificate isn't yet valid: %s", existingCert.NotBefore)
+	}
+	return nil
+}
+
+// EnforcePKI processes a spec, updating content on disk, taking action as needed.
+// Returns (TTL for PKI, error).  If an error occurs, the ttl is at best
+// a hint to the invoker as to when the next refresh is required- that said
+// the invoker should back off and try a refresh.
+func (spec *Spec) EnforcePKI(enableActions bool) error {
 
-	lifespan := time.Duration(0)
-	if !spec.Ready() {
-		log.Debugf("manager: %s isn't ready", spec)
+	updateReason := ""
+	var currentCA *x509.Certificate
+	var err error
+	if spec.renewalForced {
+		updateReason = "key"
 	} else {
-		log.Debugf("manager: %s checking lifespan", spec)
-		lifespan = spec.Lifespan()
-	}
-	log.Debugf("manager: %s has lifespan %s", spec, lifespan)
-	if lifespan <= 0 {
-		err := spec.RenewPKI()
+		currentCA, err = spec.CA.getRemoteCert()
```

Comment:
I see that we got rid of the `CompareCertificates()` function. Why do we no longer check that the remote CA cert matches the one on disk?

Reply:
We still do the comparison, it just got moved inline since 'EnforcePKI' is now dealing mostly in certificate structures (rather than raw pem).   See https://github.com/cloudflare/certmgr/blob/79caef83c1d0dfcfd6c585a2911261c0f59605d4/cert/spec.go#L430 for the path that validates whether the ondisk CA matches the fetched CA.